### DISCUSSION TITLE
Нумерация строки в зависимости от даты в поле Дата

### DISCUSSION THREAD

**Discussion by UPRLS9000** [totumonline/totum-ru-issues-and-discussions#521] [Как это сделать?] [answered]
Здравствуйте.
Подскажите, пожалуйста, как сделать правильную нумерацию строк в поле $new_n в зависимости от даты в поле Дата1.

Вот на скрине 1 показано, что добавлены строки с датами по порядку, нумерация моим кодом ставится правильно. Но если добавить новые строки с датами как в уже существующих строках или с более ранними датами, то нумерация ломается, как на скрине 2.

Скрин 1:
![1](https://github.com/totumonline/totum-ru-issues-and-discussions/assets/81652976/0eeb2064-50e6-43af-93c8-368ec2dc8158)

Скрин 2:
![2](https://github.com/totumonline/totum-ru-issues-and-discussions/assets/81652976/9b4fb930-c35e-4cc2-91e5-c6310e1b80b2)

**Comment by erpic-ru** [comment]
Не до конца понятно, что такое правильная нумерация, но вот такой код должен шпарить по порядку:
```

= :$1 + 1

1:select(table: $#ntn; field: $#nf;  order: $#nf desc; where: 'id'!= #id)
```

**Reply by UPRLS9000** [reply]
Извините, что плохо объяснил.

Нужно, чтобы в поле $new_n при добавлении новой строки в табличную часть проставлялся порядковый номер в соответствии с датой в поле Дата1.

**Пример**
есть таблица в ней есть строки, состоящие из двух столбцов Дата и Порядковый номер:
01.03.24  1
07.03.24  2
20.03.24  3

Как сделать, чтобы значения в столбце Порядковый номер пересчитывались в соответствии с датами, где самой ранней дате бы присваивался наименьший порядковый номер, а самой поздней наибольший?

**Продолжаю пример, чтобы лучше объяснить вопрос**
Добавляю в эту же таблицу строку со значением в поле дата 10.03.24, должно получиться так:
01.03.24  **1**
07.03.24  **2**
**10.03.24  3**
20.03.24  **4**

И еще нюанс, если добавляю строку с датой уже имеющейся в таблице, например 07.03.24, должно получиться так:
01.03.24  **1**
07.03.24  **2**
**07.03.24  3**
10.03.24  **4**
20.03.24  **5**

Как такое реализовать?

**Reply by erpic-ru** [reply]
Можно сильно по разному решить задачу. Например, такой код:

```
= :$1 + 1
1:select(table: $#ntn; field: $#nf; where: 'id'!= #id; where: 'data1' <= #data1;  order:  'data1' desc; order 'id' desc)
```

+ код действия при добавлении изменении/ удалении строки, который будет пересчитывать поля: 

```
a1= :reCalculate(table: $#ntn; where: 'data1' >= #data1 )
dl1= :reCalculate(table: $#ntn; where: 'data1' >= #old.data1 )

```

**Reply by UPRLS9000** [reply]
Ваш код такой результат выдает:
![4](https://github.com/totumonline/totum-ru-issues-and-discussions/assets/81652976/691d12e7-8360-4129-87b9-6a307cf8f8e2)

**Comment by ttm-alexeypolunin** [comment, marked as answer]
Для этой задачи вам потребуется техническое поле типа строка в котором надо склеить Дату и ID. Оно нужно для того, что бы обеспечить правильный порядок расчета строк.

* Ранее здесь была информация о поле сортировки, но она неверная — поле сортировки не влияет на порядок расчета строк (по возрастанию id или n) 

<img width="506" alt="Снимок экрана 2024-03-09 в 17 34 35" src="https://github.com/totumonline/totum-ru-issues-and-discussions/assets/55755565/f55aab0d-3385-4fe6-8f07-e1bf92f159a3">

Код в нем должен добавлять ведущие нули для id:

```
= : str`#date + "-" + $repeat + #id`

repeat: strRepeat(str: "0"; num: $zero)
	zero: 6 - $nl
		nl: strLength(str: #id)
```

Далее в поле с нумерацией мы привязываемся именно к этому техполю:

```
= : $prev + 1

prev: select(table: $#ntn; field: $#nf; where: 'compound' < #compound; order: 'compound' desc)
```
И where и order! очень важны в этом запросе.

Если это простая таблица, то нам еще нужен код-действия, который пересчитает строки идущие после строки вставки при вставке в середину списка:

**Добавляем в Техполе compound (так меньше перепробега!).**

С триггерами на добавление, удаление, изменение (лишнее удалить):

```
ad1=: reCalculate(table: $#ntn; where: 'compound' > #compound)

dl1=: reCalculate(table: $#ntn; where: 'compound' > #old.compound)

ch1=: reCalculate(table: $#ntn; where: 'compound' > $min)
	
	min: listMin(list: json`[#compound,#old.compound]`)
```

* вроде работает, но я прям супер-дотошно не тестировал

PS

1. Код-действия лучше поставить в техполе с датой+id, что бы избежать перепробега.
2. В Расчетных и Временных это не работает тк там другая логика доступа к данным — там другой подход!

**Reply by UPRLS9000** [reply]
Спасибо

![thank-you-gif-6](https://github.com/totumonline/totum-ru-issues-and-discussions/assets/81652976/aadf75d1-9a39-418c-930b-b029b7eee23f)

**Reply by UPRLS9000** [reply]
Этот код не работает в таблице типа: Расчетная в цикле
Помогите, если это возможно.

![5](https://github.com/totumonline/totum-ru-issues-and-discussions/assets/81652976/0a143688-c07e-46c7-9f0c-6ebb24894ff6)

**Reply by ttm-alexeypolunin** [reply]
Два момента:

1. Это ведущие нули в техническом поле.
2. В расчетных не заработает тк указание поля сортировки походу не изменяет порядок расчета строк (на самом деле не косяк, а так и должно быть!). **В расчетных другой подход — смотреть ниже!**

**Reply by UPRLS9000** [reply]
Понял, спасибо.

А коммерческая версия это облако(https://ru.totum.online/cloud)? 
После того как исправят в коммерческой версии исправления появятся и в бесплатной версии с гитхаба?

**Reply by ttm-alexeypolunin** [reply]
**Вот другой подход для этой задачи для Расчетных и Временных:**

```
= : $key[0] + 1

key: listSearch(list: $list; key: "value" = #compound)
	list: selectList(table: $#ntn; field: 'compound'; order: 'compound' asc)
```

Важно включить для таблицы параметр **Поколоночный пересчет**. Те нужно что бы колонка с compound рассчиталась до конца, когда будет вычисляться номер.

Также **в расчетной таблице не нужны никакие коды действия** для этой задачи. Тк она всегда рассчитывается целиком.

Я проверил, это работает.

* Для оптимизации лучше сделать поле в футере у compound в котором делать selectList и в listSearch брать уже через #list_in_column_futer. Это не супер-пупер ускорит, тк у БД тоже есть кэш запросов, но ускорит.

**Reply by ttm-alexeypolunin** [reply]
Внес корректировки тк это я перегрелся относительно того, что поле сортировки меняет порядок расчета — **на самом деле нет ни в расчетных ни в простых и не будет никогда это делать и так и должно быть!**
